Add event highlighting to the TidalCycles package

thgrund opened the pull request:
I implemented the event highlighting in the TidalCycles pulsar package :-) 

See this thread in the TidalCycles forum: https://club.tidalcycles.org/t/event-highlighting/4438/6?u=mrreason

ndr-brt commented:
that's really great! I'll go through a deeper review hopefully today, in the meantime I tried it, it works nicely! 
I think I found out a bug, if I evaluate a line on the very first line in the editor I get a:
```
 [GHC-83865]
    • Couldn't match expected type: Pattern String
                  with actual type: Int -> a0 -> a0
    • Probable cause: ‘deltaContext’ is applied to too few arguments
      In the first argument of ‘(-)’, namely ‘deltaContext 7’
      In the first argument of ‘s’, namely ‘(deltaContext 7 - 1 "bd")’
      In the second argument of ‘($)’, namely
        ‘s (deltaContext 7 - 1 "bd")’
```

thgrund commented:
> that's really great! I'll go through a deeper review hopefully today, in the meantime I tried it, it works nicely! I think I found out a bug, if I evaluate a line on the very first line in the editor I get a:
> 
> ```
>  [GHC-83865]
>     • Couldn't match expected type: Pattern String
>                   with actual type: Int -> a0 -> a0
>     • Probable cause: ‘deltaContext’ is applied to too few arguments
>       In the first argument of ‘(-)’, namely ‘deltaContext 7’
>       In the first argument of ‘s’, namely ‘(deltaContext 7 - 1 "bd")’
>       In the second argument of ‘($)’, namely
>         ‘s (deltaContext 7 - 1 "bd")’
> ```

I will wait for the rest of your review, before I do something, but thanks for the first bug report.

ndr-brt commented on lib/repl.js:
```diff
@@ -65,16 +66,23 @@ export default class REPL {
+
+    this.eventHighlighter.setCurrentEditor(this.editors.currentEditor());
+
+    command = this.eventHighlighter.addMetadata(command, lineNumber);
```
couldn't we pass the current editor and and make it part of the markers? that way we will be able to support multiple editor tabs out of the box, and we could remove the lines from 81 to 83

thgrund replied:
I had a look into this topic. It's absolutely no problem to pass the editors object to the `eventHighlighter` constructor. That's not a big deal. But anyway, to be able to support multiple editor tabs and duplicated tabs, we need to abuse the current `deltaContext` function a little bit. 

Here is the point: when receiving OSC messages from TidalCycles, then TidalCycles has no informations about the editor. This means, that I am not able to map the events to a text editor. But I can send at least one custom information to TidalCycles, because the `rowStart` and `rowEnd` of the `deltaContext` function will always be the same. TidalCycles is not able to treat patterns, that are longer then one row. 
As far as I understood it, in the text.management editor this was used to create a unique id: 
https://github.com/tidalcycles/Tidal/issues/1047#issuecomment-1793326603

```
Our plan is to use the "row" element of the deltaContext position as just an incrementing unique ID assigned to each mininotation string, and then using the relative offsets within that string
```

Because of the marker and decorator concept of Pulsar, this wasn't needed for me. And we could use it for whatever we want.

When we agree, using this `rowEnd` property to send custom data, then we can decide what we wanna add. Note: every additional information will be send with every osc message.

**Option 1: TextEditor id** 

Example in the console: 
```
atom.workspace.getTextEditors()[0].id 
// example value 178
```

- Smaller id value
- Can only be used to identify a specific text editor

**Option 2: TextBuffer id** 

Example in the console: 
```
atom.workspace.getTextEditors()[0].getBuffer().id 
// example value a7c301dda95683ce1e20dd18891e4461
```
- Bigger id value
- Can be used to identify all text editors with the same buffer


WDYT @ndr-brt ?

ndr-brt replied:
I definitely think we should assign an id, I see the second option as the best one

ndr-brt commented on lib/repl.js:
```diff
@@ -65,16 +66,23 @@ export default class REPL {
 
-  tidalSendLine(command) {
+  tidalSendLine(command, lineNumber) {
+    if (!this.editors.currentIsTidal()) return;
```
this should be necessary as there's no way to send a tidal line on a non-tidal editor. Am I correct?

thgrund commented on lib/osc-loader.js:
```diff
@@ -3,28 +3,34 @@
 
+            const resultDict = this.asDictionary(resultMap.args)
+
+            if (resultMap.address === "/pulsar/event-highlighting/add-style") {
```
Valid points! I wonder how we wanna make it configurable. Maybe we should make it feature wide. For example instead of OSC evaluation we could rename it to remote control (eval, cursor jump by search etc) and event highlighting. Each feature, that needs it, has their own OSC configurations. 

How I register it is something that I will wrap my head around. We could create a simple map with the feature name as the key and register it (like you said)

WDYT?

ndr-brt replied:
yes, agree, the design could be:
- `OscServer`: given an ip and port sets up an osc server, on which one or more `OscLoader` (that could be renamed `OscListener` or `RemoteControl`) can be registered
- `OscLoader`/`OscListener`/`RemoteControl` (I think `OscListener` is my favorite): is a pair of an address and a function that get the `resultDict` as input.

I'd also say that I don't see the need to have multiple `OscServer`s running. A single `OscServer` could be instantiated at the plugin startup and passed to the components that requires to register a listener on it, then the server will be started only if at least one listener is registered.

This looks definitely tidier. I can take care of this refactor on a separated PR so you'll just need to rebase on these changes and focus on the highlight part

thgrund commented on lib/osc-server.js:
```diff
@@ -11,8 +10,8 @@ export default class OscServer {
-        this.port =  atom.config.get(oscPortProperty);
+    constructor(consoleView, port) {
+        this.port =  port;
         this.ip =  atom.config.get(oscIpProperty);
```
See https://github.com/tidalcycles/pulsar-tidalcycles/pull/213#discussion_r2205198569

thgrund commented:
> > that's really great! I'll go through a deeper review hopefully today, in the meantime I tried it, it works nicely! I think I found out a bug, if I evaluate a line on the very first line in the editor I get a:
> > ```
> >  [GHC-83865]
> >     • Couldn't match expected type: Pattern String
> >                   with actual type: Int -> a0 -> a0
> >     • Probable cause: ‘deltaContext’ is applied to too few arguments
> >       In the first argument of ‘(-)’, namely ‘deltaContext 7’
> >       In the first argument of ‘s’, namely ‘(deltaContext 7 - 1 "bd")’
> >       In the second argument of ‘($)’, namely
> >         ‘s (deltaContext 7 - 1 "bd")’
> > ```
> 
> I will wait for the rest of your review, before I do something, but thanks for the first bug report.

Done @ndr-brt

thgrund commented:
@ndr-brt thanks for this improvement: https://github.com/tidalcycles/pulsar-tidalcycles/pull/218
I refactored the code and applied the new osc server handling. I needed to extend it a little bit:

1. When receiving osc bundles, then we need to apply the registered functions to each element. The event highlighting osc messages are bundled message, but they only contain one element. Unfortunately this is counter intuitive.
2. The event highlighting osc messages do not send key/value pairs. But I would prefer to stick with that approach. So I extended the subscriber function to be able to receive an optional transform function. This ensures, that we can internally use the same structure.

The subscriber functions are defined within the event highlighting class. And I extracted all osc calls and moved it to the main (tidalcycles.js) file.

thgrund commented on lib/tidalcycles.js:
```diff
@@ -37,12 +38,22 @@ export default {
         this.ghc = new Ghc(this.consoleView);
         this.ghc.init();
+
+        const oscEventHighligthingServer = new OscServer(this.consoleView, atom.config.get('tidalcycles.eventHighlighting.ip'), atom.config.get('tidalcycles.eventHighlighting.port'));
```
I get your point, but I wanted to be able to toggle the feature on the fly. When I restrict it here, then you need to launch TidalCycles with this feature enabled. And the feature itself is off by default. WDYT?

ndr-brt replied:
generally all the features are enabled at the plugin startup (see oscEval), so I don't see many advantages to go differently, I can image that a person using the plugin want that feature or not, being togglable during execution doesn't look like a necessary use case

thgrund replied:
Moved it - Done

thgrund commented on lib/osc-loader.js:
```diff
@@ -0,0 +1,71 @@
+const oscPortProperty = 'tidalcycles.oscEval.port';
+const oscAddressProperty = 'tidalcycles.oscEval.address';
+
+export default class OscLoader extends EventEmitter {
```
Done

thgrund commented on lib/osc-server.js:
```diff
@@ -32,14 +33,18 @@ export default class OscServer {
-            let subscriber = this.subscribers.get(oscMessage.address);
-            if (subscriber) {
-                subscriber(this.#asDictionary(oscMessage.args))
+            if (oscmessage.oscType === "bundle") {
```
I think we need to think about tests in general. What would be your inteded strategy? I can add a test for this specific case. But I assume you would like to see, that the event highlighter is tested as well?

thgrund replied:
Done - I needed to change the OSC-Server testsuite, because it didn't tested anything yet. It covers a single message and a bundle scenario.

thgrund commented on lib/osc-server.js:
```diff
@@ -52,6 +57,20 @@ export default class OscServer {
     }
 
+    handleOscMessage(address, args) {
+       let subscriber = this.subscribers.get(address);
```
I really like that approach! Done

ndr-brt commented on lib/event-highlighter.js:
```diff
@@ -0,0 +1,318 @@
+    }
+  }
+
+  // ----------------------------------------------------------------------
+  // Public API
+  // ----------------------------------------------------------------------
```
this kind of comments are quite misleading, because there's some public api above (init, destroy), but also below the `Private Helpers` block, I'd just remove these blocks and put the public methods on top and the private on bottom.

thgrund replied:
Done

ndr-brt commented on lib/event-highlighter.js:
```diff
@@ -0,0 +1,318 @@
+'use babel'
+
+import { Range, Point, CompositeDisposable } from "atom"; // eslint-disable-line no-unused-vars
+import path from "path"; // eslint-disable-line no-unused-vars
+const osc = require('osc-min');
```
these lines can be deleted

thgrund replied:
Done - I really have poor dx on my private machine. This is most likely, because I use Pulsar for editing the Pulsar package :-D Maybe I will think about optimizing this.

thgrund commented on lib/event-highlighter.js:
```diff
@@ -0,0 +1,318 @@
+    `);
+  }
+
+  handleStyleMessage(payload) {
```
Unfortunately, I realized this today as well. The point here is, that I need add a new server for this. The idea is, that you can send CSS properties over OSC to change the styles from outside (on the fly). My use case is, that I use the Orbit colors that I defined in the SuperDirtMixer. This became unused, because of the refactorisations. Will change it.

ndr-brt replied:
I will propose to add this feature in a separate PR eventually, once the highlight is fully completed

thgrund replied:
Done

thgrund commented on lib/event-highlighter.js:
```diff
@@ -0,0 +1,318 @@
+
+    // Replace quoted segments with `(deltaContext …)` unless they belong to
+    // known control‑pattern contexts.
+    return line.replace(/"([^"]*)"/g, (match, content, offset) => {
+      const before = line.slice(0, offset);
+      if (/numerals\s*=.*$|p\s.*$/.test(before)) return match; // ignore control‑patterns
```
Hmmm, I would keep it here, because it better fits the scope. But it should be tested in any case. 

I agree that I will create helper functions that can be tested in isolation. With this, we would introduce a more descriptive meaning for these regex statements as well. Another option would be to treat regex statements like magic numbers. But anyway - it's not needed when we create helper functions (an it makes it testable)

But in any case the regex in line 80 is something, that I really hate in that implementation. It behaves like a blacklist of commands, that shall be ignored. We treat every string as a ControlPattern. With a list of exceptions, you can allow functions that shouldn't receive the `deltaContext` function. This was one of the reasons, why I wanted to toggle the feature on the fly. I.e. when you create a custom stream in you ghci session like `p "myStream"`, then this will throw an error, because the value of p is an `ID`.

thgrund replied:
Oooookay...I wrapped my head around this a little bit more. And I would like to introduce a helper file that I call "line-processor.js". I would start with just creating a bunch of static functions inside of a class `LineProcess` and create the tests for them.

thgrund commented:
@ndr-brt you can give this PR another round of review. I resolved everything you came up so far.

ndr-brt reviewed (approved):
sorry for taking too long, LGTM and I will roll out a new version so people can start testing it appropriately! thanks for the huge effort!
